nix/eval-machine-info.nix

```diff
@@ -11,104 +11,170 @@
+                    default = {};
+                    type = types.attrsOf (types.submodule (r:{
+                      _module.args = {
+                        inherit pkgs uuid;
+                        resources = r.config;
+                        # inherit nodes, essentially
+                        nodes =
+                          lib.mapAttrs
+                            (nodeName: node:
+                              lib.mapAttrs
+                                (key: lib.warn "Resource ${r.name} accesses nodes.${nodeName}.${key}, which is deprecated. Use the equivalent option instead: nodes.${nodeName}.${newOpt key}.")
+                                config.nodes.${nodeName})
+                            config.nodes;
+                      };
+                      imports = [
+                        mainModule
+                        deploymentInfoModule
+                        ./resource.nix
+                      ];
+                    }));
+                  };
+                in {
+                  sshKeyPairs = resOpt ./ssh-keypair.nix;
+                  commandOutput = resOpt ./command-output.nix;
+                };
+                config = {
+                  machines = config.nodes;
+                  _module.check = false;
+                };
+              }] ++ pluginResources;
+              specialArgs = {
+                inherit evalResources resourcesByType;
+                inherit (lib) zipAttrs;
+              };
+            };
+          };
+          # Compute the definitions of the machines.
+          nodes = mkOption {
+            description = "The NixOS configurations for the nodes in the network.";
+            type = types.attrsOf (import "${nixpkgs}/nixos/lib/eval-config.nix" {
+              inherit lib system;
+              specialArgs = {
+                inherit uuid deploymentName;
+                inherit (config) nodes resources;
+              } // config.network.nodesExtraArgs;
+              modules = # Make NixOps's deployment.* options available.
+                pluginOptions ++
+                [
+                  ./options.nix
+                  ./resource.nix
+                  deploymentInfoModule
+                  config.defaults
+                  ({ name, ... }: rec{
+                    _file = ./eval-machine-info.nix;
+                    key = _file;
+                    # Provide a default hostname and deployment target equal
+                    # to the attribute name of the machine in the model.
+                    networking.hostName = lib.mkOverride 900 name;
+                    deployment.targetHost = lib.mkOverride 900 name;
+                    environment.checkConfigurationOptions = lib.mkOverride 900 checkConfigurationOptions;
+                    nixpkgs.system = lib.mkDefault system;
+                  })
+                ];
+            }).type;
+          };
+          defaults = mkOption {
+            type = types.anything;
```

Comment:
`anything` would try to merge things on its own when multiple `defaults` values are defined. Instead we ought to defer this to the module system.
Something like https://github.com/NixOS/nixpkgs/pull/152785 or perhaps better, a type that returns a `mkMerge`.

Reply:
Let's keep it simple. `types.raw` should be fine in almost all cases. It seems unlikely that the module system would touch those values too early, and impossible when they're in function bodies, which is very often the case.

Reply:
Here's a new type that accepts single modules and returns them as a list suitable for `imports` or `modules`. https://github.com/NixOS/nixpkgs/pull/163617